### PR TITLE
More refactor of attentions  & Add MQA GQA 

### PR DESCRIPTION
**Pull request by hanzlfs**
1. More MHA refactor (remove attention norm)
2. support MQA GQA(for Llama 70B)

### CODE REVIEW

File: MaxText/tests/attention_test.py [hunk 1]
```diff
@@ -1,110 +1,402 @@
...
+    )
+
+    self.assertTrue(
+        jax.numpy.allclose(
+            flash_output, mha_output, rtol=1e-01, atol=1e-01, equal_nan=False
+        )
+    )
+
+  def test_multihead_attention(self):
+    """Test MultiHeadDotProductAttention."""
```
Comment (rwitten) [inline MaxText/tests/attention_test.py:155]: I find the reimplementation kind of weird! Shouldn't we just test that MHA == FLASH_MHA?

File: MaxText/tests/attention_test.py [hunk 2]
```diff
@@ -1,110 +1,402 @@
...
+    weights = jax.nn.softmax(logits, axis=-1)
+    combined = jnp.einsum('bnts,bsnd->btnd', weights, value)
+    expected = jnp.einsum('btnd,ndf->btf', combined, out_kernel)
+
+    out = np.asarray(out).astype(np.float16)
+    expected = np.asarray(expected).astype(np.float16)
+    np.testing.assert_allclose(out, expected, atol=1e-6)
+
+  def test_multiquery_attention(self):
+    """Test MultiQueryDotProductAttention."""
```
Comment (rwitten) [inline MaxText/tests/attention_test.py:235]: Again I don't love the reimplementation. Shouldn't we test that MQA is the same as MHA (with queries replicated)?

File: MaxText/tests/attention_test.py [hunk 3]
```diff
@@ -1,110 +1,402 @@
...
+    weights = jax.nn.softmax(logits, axis=-1)
+    combined = jnp.einsum('bnts,bsd->btnd', weights, value)
+    expected = jnp.einsum('btnd,ndf->btf', combined, out_kernel)
+
+    out = np.asarray(out).astype(np.float16)
+    expected = np.asarray(expected).astype(np.float16)
+    np.testing.assert_allclose(out, expected, atol=1e-6)
+
+  def test_groupedquery_attention(self):
+    """Test GroupedQueryDotProductAttention."""
```
Comment (rwitten) [inline MaxText/tests/attention_test.py:317]: Again, I would just test that GQA recovers MHA where the keys replicated.

**Review by rwitten — CHANGES_REQUESTED** [review]
LOL you've made me OK with the code but I still find the tests annoying. The tests are all a "reimplementation" which will be annoying when the code inevitably changes. Wouldn't it be better to tests MHA == FlashMHA and then GQA and MQA use MHA for their tests (by replicating their KV caches?)

File: MaxText/tests/attention_test.py (hunk 1 above)
Comment (hanzlfs) [reply MaxText/tests/attention_test.py:155]: Then how we know MHA or Flash MHA is correct?

Comment (hanzlfs) [reply MaxText/tests/attention_test.py:155]: Who is our ground truth then?

File: MaxText/tests/attention_test.py (hunk 2 above)
Comment (hanzlfs) [reply MaxText/tests/attention_test.py:235]: Stacking them into a method, is not readable.

File: MaxText/tests/attention_test.py (hunk 3 above)
Comment (hanzlfs) [reply MaxText/tests/attention_test.py:317]: same as above

File: MaxText/tests/attention_test.py (hunk 2 above)
Comment (hanzlfs) [reply MaxText/tests/attention_test.py]: Refactoring this to equivalent test between GQA = MHA when num_kv_head = num_head

**Review by rwitten — APPROVED** [review]
let's go! :-)

**Pull request merged**